File: solutions/fsharp/protein-translation/1/ProteinTranslation.fs
module ProteinTranslation

let (|Prefix|_|) (pattern: string) (str: string) =
    if str.StartsWith(pattern) then
        let tail = str.Substring(pattern.Length)
        Some(tail)
    else
        None

let proteins rna =
    let rec matcher str acc =
        match str with
        | Prefix "AUG" tail -> matcher tail ("Methionine" :: acc)
        | Prefix "UUU" tail -> matcher tail ("Phenylalanine" :: acc)
        | Prefix "UUC" tail -> matcher tail ("Phenylalanine" :: acc)
        | Prefix "UUA" tail -> matcher tail ("Leucine" :: acc)
        | Prefix "UUG" tail -> matcher tail ("Leucine" :: acc)
        | Prefix "UCU" tail -> matcher tail ("Serine" :: acc)
        | Prefix "UCC" tail -> matcher tail ("Serine" :: acc)
        | Prefix "UCA" tail -> matcher tail ("Serine" :: acc)
        | Prefix "UCG" tail -> matcher tail ("Serine" :: acc)
        | Prefix "UAU" tail -> matcher tail ("Tyrosine" :: acc)
        | Prefix "UAC" tail -> matcher tail ("Tyrosine" :: acc)
        | Prefix "UGU" tail -> matcher tail ("Cysteine" :: acc)
        | Prefix "UGC" tail -> matcher tail ("Cysteine" :: acc)
        | Prefix "UGG" tail -> matcher tail ("Tryptophan" :: acc)
        | _ -> acc

    matcher rna [] |> List.rev

File: solutions/csharp/protein-translation/1/ProteinTranslation.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class ProteinTranslation
{
    public static IEnumerable<T> ForEach<T>(this IEnumerable<T> source, Action<T> action)
    {
	foreach (var item in source)
	    action(item);
	return source;
    }
    public static string[] Proteins(string strand)
    {
	return  strand.Chunk(3)
	    .Select(chk => new string(chk))
	    .Select(str => str switch {
		"AUG" => "Methionine",
		"UUU" or "UUC" => "Phenylalanine",
		"UUA" or "UUG" => "Leucine",
		"UCU" or "UCC" or "UCA" or "UCG" => "Serine",
		"UAU" or "UAC" => "Tyrosine",
		"UGU" or "UGC" => "Cysteine",
		"UGG" => "Tryptophan",
		"UAA" or "UAG" or "UGA" => "STOP",
		_ => "ERROR"
            }).TakeWhile(str => !str.Equals("STOP"))
	    .ForEach((str) => {
	        if (str.Equals("Error"))
                    throw new Exception($"Invalid condon {str}");
	    }).ToArray();
    }
}
